Support PR creation on Bitbucket

iinuwa opened the pull request:
This should fix #2919, but depends on #2918.

This gets us most of the way, but I'd appreciate some help polishing it off.

I was able to successfully create a PR with Bitbucket Cloud by modifying dependabot-script (cf. dependabot/dependabot-script#514), but I haven't added any automated tests because I don't really know where to start.

I also had a few questions as I was going through the code. I'll add review comments below.

iinuwa commented on common/lib/dependabot/clients/bitbucket.rb:
```diff
@@ -12,12 +12,28 @@ class Unauthorized < StandardError; end
 
-      def initialize(credentials:)
+      # FIXME: I don't know if changing the constructor here is safe
+      def initialize(credentials:, source: nil)
```
Is this save to add an extra parameter here? (I hadn't done any Ruby before this, so I don't know about this.) It seems to match how the other constructors work.

thepwagner replied:
Question: what is using this new `source` field? I skimmed and don't see any references.

thepwagner replied:
It should be, since you're providing a default value.

iinuwa replied:
I didn't add any references to it since I wasn't sure whether I should prefer a `repo` parameter passed to each method, or to use the instance variable. I went ahead and removed the parameter since I don't think `source` applies to Bitbucket anyway.

(There isn't really an alternative location for Bitbucket. Support for Bitbucket Server is going away in Feb 2024 and the ability to configure custom domain names for Bitbucket Cloud isn't on the Atlassian roadmap to be implemented any time soon, so `bitbucket.org`/`api.bitbucket.org` are the only endpoints that we need to consider—unless someone who can test Bitbucket Server requests the feature.)

iinuwa commented on common/lib/dependabot/clients/bitbucket.rb:
```diff
@@ -52,6 +68,92 @@ def fetch_file_contents(repo, commit, path)
         response.body
       end
 
+      def commits(repo, branch_name = nil)
```
I copied this most of the new methods from the Azure client, but I had to add the `repo` parameter to all of them.

Is this supposed to be using `source.repo` internally for all these parameters?

iinuwa commented on common/lib/dependabot/clients/bitbucket.rb:
```diff
@@ -52,6 +68,92 @@ def fetch_file_contents(repo, commit, path)
 
+      def commits(repo, branch_name = nil)
+        commits_path = "#{repo}/commits/" + branch_name.to_s
+        commits_path += "?pagelen=100"
```
This only returns the last 100 commits; should this go paginate through all commits instead?

thepwagner replied:
I suspect other clients do, but 100 commit seems like it might be "enough".
If it's working for you, works for me...

iinuwa replied:
Meh. I went ahead and implemented pagination for both commits and pull requests so someone else doesn't have to do it later.

iinuwa commented on common/lib/dependabot/clients/bitbucket.rb:
```diff
@@ -52,6 +68,92 @@ def fetch_file_contents(repo, commit, path)
+          next_page = result.fetch("next")
+        end
+
+        pull_requests unless source_branch && target_branch
```
Are we supposed to bail if either `source_branch` or `target_branch` is `nil`? Or should we let the provider determine defaults for us if not specified in the client?

iinuwa commented on common/lib/dependabot/clients/bitbucket.rb:
```diff
@@ -52,6 +68,92 @@ def fetch_file_contents(repo, commit, path)
+              name: target_branch
+            }
+          },
+          description: pr_description,
```
Bitbucket doesn't like the HTML tags in the generated description; does that need to be fixed in this PR before merging?

iinuwa commented on common/lib/dependabot/pull_request_creator/bitbucket.rb:
```diff
@@ -0,0 +1,98 @@
+      def create
+        return if branch_exists? && pull_request_exists?
+
+        # FIXME: Copied from Azure, but not verified whether this is true
+        # For Bitbucket we create or update a branch in the same request as creating
+        # a commit (so we don't need create or update branch logic here)
+        create_commit
```
As the comment says, this logic needs to be reviewed.

iinuwa replied:
Updating PRs is a [different endpoint](https://developer.atlassian.com/bitbucket/api/2/reference/resource/repositories/%7Bworkspace%7D/%7Brepo_slug%7D/pullrequests/%7Bpull_request_id%7D#put), so we either need to throw an error when attempting to update, or add the endpoint reference.

iinuwa commented on common/lib/dependabot/pull_request_creator/bitbucket.rb:
```diff
@@ -0,0 +1,98 @@
+
+      private
+
+      # FIXME: I don't know what the "for_source" methods are for
+      def bitbucket_client_for_source
```
I mentioned this above, but I'm not really sure what the `for_source` methods are for. Could someone explain this a little?

thepwagner replied:
I think their main goal is to instantiate the client once and only once (e.g. the `@bitbucket_client_for_source ||=` line is the key).

I'm not sure about the naming - AFAICT there is  only ever one `source` (and therefore one `bitbucket_client`).

iinuwa replied:
Gotcha. I went ahead and removed the comment. I agree with the "only ever one source" part. 

> AFAICT there is only ever one `source`

Yeah, I agree. It seems to be the convention for the other clients, though, so I'll defer to you on the name.

thepwagner commented on common/lib/dependabot/clients/bitbucket.rb:
```diff
@@ -52,6 +68,92 @@ def fetch_file_contents(repo, commit, path)
       end
 
+      def commits(repo, branch_name = nil)
+        commits_path = "#{repo}/commits/" + branch_name.to_s
+        commits_path += "?pagelen=100"
```
```suggestion
        commits_path = "#{repo}/commits/#{branch_name}?pagelen=100"
```

Idea: simplify with string interpolation.

thepwagner commented on common/lib/dependabot/clients/bitbucket.rb:
```diff
@@ -52,6 +68,92 @@ def fetch_file_contents(repo, commit, path)
+        }
+
+        files.each do |file|
+          absolute_path = "/" + file.name unless file.name.start_with?("/")
```
Concern: what value should `absolute_path` take if `file.name.start_with?("/") == true`?
It might be `nil` currently, which would mean multiple absolute paths conflict.

iinuwa replied:
Ah, that was a mistake; I've corrected that.

thepwagner commented on common/lib/dependabot/clients/bitbucket.rb:
```diff
@@ -52,6 +68,92 @@ def fetch_file_contents(repo, commit, path)
+
+        body = encode_form_parameters(parameters)
+
+        commit_path = "#{repo}/src"
```
Xref https://developer.atlassian.com/bitbucket/api/2/reference/resource/repositories/%7Bworkspace%7D/%7Brepo_slug%7D/src#post

Not worth an inline comment, just sharing for fellow reviewers.

thepwagner commented on common/lib/dependabot/pull_request_creator/bitbucket.rb:
```diff
@@ -0,0 +1,98 @@
+          source.branch || default_branch,
+          pr_description,
+          # TODO: Bitbucket doesn't yet support labels. Cf. https://jira.atlassian.com/browse/BCLOUD-11976
+          # labeler.labels_for_pr,
```
```suggestion
          nil,
```

Comment SGTM, but since `labels` is a positional arg it needs to be provided so that `work_item` is passed correctly.

thepwagner commented on common/lib/dependabot/pull_request_creator/bitbucket.rb:
```diff
@@ -0,0 +1,98 @@
+      end
+
+      def branch_exists?
+        bitbucket_client_for_source.branch(source.repo, branch_name)
```
Nit: it's weird that we passed `source` to the constructor for `Dependabot::Clients::Bitbucket.for_source`, but still have to pass attributes like `source.repo` to individual methods.

If `source` isn't used by `Dependabot::Clients::Bitbucket` it may be cleanest to remove.

thepwagner commented on common/lib/dependabot/pull_request_creator/pr_name_prefixer.rb:
```diff
@@ -307,6 +309,19 @@ def recent_azure_commit_messages
           map(&:strip)
       end
 
+      # FIXME: Should this paginate through all commits, or is 100 enough?
```
```suggestion
```

I think this is expressed in the `commits()` function already.

thepwagner commented on common/lib/dependabot/pull_request_creator/pr_name_prefixer.rb:
```diff
@@ -307,6 +309,19 @@ def recent_azure_commit_messages
+
+        @recent_bitbucket_commit_messages.
+          reject { |c| bitbucket_commit_author_email(c) == dependabot_email }.
+          reject { |c| c.fetch("message")&.start_with?("Merge") }.
+          map { |c| c.fetch("message") }.
+          compact.
```
```suggestion
          map { |c| c.fetch("message") }.
          compact.
          reject { |m| m.start_with?("Merge") }.
```

Idea: reduce to commits with a `message` before filtering on the message contents.

iinuwa replied:
Done, with a modification.

thepwagner reviewed (commented):
LGTM, thanks for this!

I'm also confused about how `source` provided to these clients' constructor should interact with `source.repo` required by methods. This long predates me, @feelepxyz or @hmarr is this part of an ongoing refactor? Which should be preferred?

iinuwa commented on common/lib/dependabot/clients/bitbucket.rb:
```diff
@@ -12,12 +12,28 @@ class Unauthorized < StandardError; end
+      # Constructor methods #
+      #######################
+
+      def self.for_source(source:, credentials:)
```
Yeah, I noticed that when I tried to remove the `for_source` method. It looks like all the clients do the same thing.

iinuwa commented:
@thepwagner, I think I've addressed the suggestions you've made so far. I've put them in individual commits in case that's helpful for reviewing. I can squash and rebase after approval and before merging to clean up the history.

thepwagner reviewed (approved):
Thanks for this!
LGTM: I've regression tested with the `dry-run.rb` script in this repository.

dependabot/dependabot-script isn't supported by GitHub, so I haven't tested this there - follow-up PRs are welcome if more is needed.

meladRaouf commented:
I can't use dependabot with bitcucket I got this error 
labeler.rb:241:in `labels': Unsupported provider bitbucket (RuntimeError)

is there anyway to skip labels in bitbucket
